Asio-Server-and-Client/queue.cpp

```diff
@@ -0,0 +1,118 @@
+#include <cmath>
+#include <functional>
+#include <iostream>
+#include <mutex>
+#include <stdexcept>
+#include <thread>
+
+template<typename T, uint64_t SIZE = 4096, uint64_t MAX_SPIN_ON_BUSY = 40000000>
+class ConcurrentQueue {
+private:
+    static constexpr unsigned Log2(unsigned n, unsigned p = 0) {
+        return (n <= 1) ? p : Log2(n / 2, p + 1);
+    }
```

Comment:
I think we can rewrite this function to have only one argument
```suggestion
    static constexpr unsigned Log2(unsigned n) {
        return (n < 2) ? 0 : 1 + Log2(n / 2);
    }
```